lib/puppet/provider/nameservice.rb

```diff
@@ -288,5 +288,25 @@ def set(param, value)
       raise Puppet::Error, "Could not set #{param} on #{@resource.class.name}[#{@resource.name}]: #{detail}", detail.backtrace
     end
   end
+
+  # From overriding Puppet::Property#insync?
+  # Ruby Etc::getpwnam < 2.1.0 always returns a struct with binary encoded
+  # string values, and >= 2.1.0 will return binary encoded strings for values
+  # incompatible with current locale charset. Convert "should" value to binary
+  # if "current" value is, to avoid unnecessary property syncs and concatenation
+  # of strings with incompatible encodings. (PUP-6777)
+  #
+  # * "should" parameter will always be a string with one possible (strictly
+  # speaking) exception, which is the boolean value true. In current semantics,
+  # this would convert to a string. For strict backwards-compatibility, convert
+  # it to a string if this is the case.
+  #
+  # return basic string comparison after encoding (same as
+  # Puppet::Property#property_matches)
+  def comments_insync?(current, should)
+    should = should.to_s if should == true
+    should.force_encoding(Encoding::ASCII_8BIT) if current.encoding == Encoding::ASCII_8BIT
```

Comment:
I'm confident given the constraints around this provider that we really only need to handle the case where we're dealing with a binary-encoded 'is', but another possibility here would be to do `should.force_encoding(current.encoding)` - feels riskier though but might be more..thorough.